blueprint/core/blueprint_ViewManager.h

```diff
@@ -0,0 +1,87 @@
+*/
+
+#pragma once
+
+#include <map>
+
+#include "blueprint_View.h"
+#include "blueprint_ShadowView.h"
+
+namespace blueprint
+{
+    //TODO: Document
+    class ViewManager : public juce::ComponentListener
+    {
+    public:
+        //==============================================================================
+        // We allow registering arbitrary view types with the React context by way of
+        // a "ViewFactory" here which is a user-defined function that produces a View
+        // and a corresponding ShadowView.
+        typedef std::pair<std::unique_ptr<View>, std::unique_ptr<ShadowView>> ViewPair;
+        typedef std::function<ViewPair()> ViewFactory;
+
+        //==============================================================================
+
+        //TODO: Document
+        explicit ViewManager(View* rootView);
+        ~ViewManager();
+
+        /** Registers a new dynamic view type and its associated factory. */
+        void registerViewType(const juce::String& typeId, ViewFactory f);
+
+        /** Creates a new view instance and registers it with the view table. */
+        ViewId createViewInstance(const juce::String& viewType);
+
+        /** Creates a new text view instance and registers it with the view table. */
+        ViewId createTextViewInstance(const juce::String& value);
+
+        //TODO: Document
+        void setViewProperty(ViewId viewId, const juce::Identifier& name, const juce::var& value);
+
+        //TODO: Document
+        void setRawTextValue(ViewId viewId, const juce::String& value);
+
+        //TODO: Document
+        void addChild(ViewId parentId, ViewId childId, int index = -1);
+
+        //TODO: Document
+        void removeChild (ViewId parentId, ViewId childId);
+
+        //==============================================================================
+        // juce::ComponentListener
+        void componentMovedOrResized (juce::Component& component, bool wasMoved, bool wasResized) override;
+
+        //==============================================================================
+    private:
+        void enumerateChildViewIds (std::vector<ViewId>& ids, View* v);
+
+        /** Recursively computes the shadow tree layout on the root ShadowView, then traverses the tree
+            flushing new layout bounds to the associated view components.
+         */
+        void performRootShadowTreeLayout();
+
+        /** Returns a pointer pair to the view associated to the given id. */
+        //TODO: Private for now as only used internally but any desire for this to be made public?
+        //      so classes like ReactApplicationRoot can fetch a View?
+        std::pair<View*, ShadowView*> getViewHandle (ViewId viewId);
+
+        /** Walks the view table, returning the first view with a `refId`
+         *  whose value equals the provided id.
+         */
+         //TODO: Currently unused. Keep?
+        View* getViewByRefId (const juce::Identifier& refId);
```

Comment:
Yea let's keep this one, it's useful for client code. For example, in this plugin WOV by K-Devices (https://k-devices.com/products/wov/), which is written in blueprint, those lines drawn between the knobs are done in the PluginEditor.h `paint()` method, and in that paint routine the editor asks the appRoot "hey do you have a view tagged by the refId `gainSlider`"? And with the `View*` the paint() routine is able to identify the exact location of that view and draw a line to it.

Then from the React code we can just label that view with the refId via props:

```
<Slider refId="gainSlider" />
```

It's neat when you want to kind of interoperate between react and native paint routines. Of course that was before we started talking about the Canvas and drawing stuff :) But I think this still makes sense here for now

Reply:
Ahh nice. I was actually wondering how obtaining views by the equivalent of a component I'd should work for a separate problem in my app. Cool, makes perfect sense. Will kill the TODO